subsys/power/device_pm.c

```diff
@@ -0,0 +1,226 @@
+struct k_work_q devpm_workq;
+
+/*
+ * Device PM worker to handle the Async PM requests
+ * from device_pm_get/device_pm_put API's. The worker
+ * will call the device_set_power_state() to put the
+ * device in desired power state and notify the caller
+ * through k_alert mechanism.
+ */
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, dwork);
+	struct device *dev = device_get_binding(pm->name);
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+
+	/* Incase of suspend request, make sure the device is not busy */
+	if (device_busy_check(dev) && (dev->pm->request
+				== DEVICE_PM_SUSPEND_STATE)) {
+		LOG_ERR("dev_pm: device busy!!\n");
+		dev->pm->status =  DEVICE_PM_REQ_STAT_ERROR;
+		goto out;
+	}
+
+	ret = device_set_power_state(dev, dev->pm->request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+					dev->pm->request, ret);
+		dev->pm->status = DEVICE_PM_REQ_STAT_ERROR;
+	} else {
+		dev->pm->status = DEVICE_PM_REQ_STAT_DONE;
+	}
+out:
+	/* Notify the driver through alert */
+	if (dev->pm->alert.handler) {
+		k_alert_send(&dev->pm->alert);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_set_state(struct device *dev,
+			       u32_t target_state, u32_t pm_flags)
+{
+	int ret;
+	u32_t current_state;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto out;
+	}
+
+	/* Return if current state is same as target state */
+	if (current_state == target_state) {
+		goto out;
+	}
+
+	/* Cancel already queued work */
+	if (dev->pm->status == DEVICE_PM_REQ_STAT_INIT) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->status = DEVICE_PM_REQ_STAT_CANCEL;
+		/* Notify the driver through alert */
+		if (dev->pm->alert.handler) {
+			k_alert_send(&dev->pm->alert);
+		}
+	}
+
+	/* Schedule the worker incase of async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		dev->pm->request = target_state;
+		dev->pm->status = DEVICE_PM_REQ_STAT_INIT;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+	} else {
+		ret = device_set_power_state(dev, target_state);
+		if (ret) {
+			LOG_ERR("Set power state(%d) error(%d)\n",
+							target_state, ret);
+		}
+	}
+
+out:
+	return ret;
+}
+
+static int device_pm_handle_get(struct device *dev, u32_t pm_flags)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		ret = -ERANGE;
+		goto get_out;
+	}
+	dev->pm->usage++;
+
+	if (!dev->pm->enable) {
+		ret = 0;
+		goto get_out;
+	}
+
+	ret = device_pm_set_state(dev,
```

Comment:
I think that we should request new pm state only if usage count crosses 0/1 boundary.

Reply:
Ok.

Reply:
I am not getting this. IMHO we have to track only 2 things: What state transition is executing and what is requested state. This gives us 4-state FSM:
- DEVICE_ACTIVE:
- DEVICE_SUSPENDING:
- DEVICE_WAKING_UP:
- DEVICE_SHUTTING_DOWN:

Depending on the request and state, you can easily make the decision what to do and exit form the sync request when given state was reached.

Reply:
Ok got it. I will add the FSM logic to distinguish the states.

Reply:
As we wanted to avoid longer wait on async calls, I introduced separate lock for handling the power state transitions and during this time, the worker will not hold the normal lock. Due to this we may not be able to implement the above logic. I will update the PR please have a look.